packages/stryker-karma-runner/src/KarmaTestRunner.ts

```diff
@@ -1,235 +1,122 @@
-    autoWatch: false,
-    // Don't stop after first run
-    singleRun: false,
-    // Never detach, always run in this same process (is already a separate process)
-    detached: false
-  };
-  return Object.freeze(config);
-})();
-
-function defaultOptions(): Readonly<ConfigOptions> {
-  return Object.freeze({
-    browsers: ['PhantomJS'],
-    frameworks: ['jasmine'],
-  });
-}
-
 export default class KarmaTestRunner implements TestRunner {
-
   private log = log4js.getLogger(KarmaTestRunner.name);
-  private server: karma.Server;
-  private serverStartedPromise: Promise<void>;
   private currentTestResults: TestResult[];
   private currentErrorMessages: string[];
   private currentCoverageReport?: CoverageCollection | CoveragePerTestResult;
-  private currentRunResult: karma.TestResults;
-  private readonly testHooksMiddleware = new TestHooksMiddleware();
+  private currentRunStatus: RunStatus;
+  private readonly testHooksMiddleware = TestHooksMiddleware.instance;
+  private readonly starter: ProjectStarter;
 
   constructor(private options: RunnerOptions) {
     setGlobalLogLevel(options.strykerOptions.logLevel || 'info');
-    let karmaConfig = this.readConfig(options);
-    karmaConfig = this.configureTestRunner(karmaConfig);
-    karmaConfig = this.configureCoverageIfEnabled(karmaConfig);
-    karmaConfig = this.configureProperties(karmaConfig);
-    karmaConfig = this.configureTestHooksMiddleware(karmaConfig);
-
-    this.log.debug(`using config ${JSON.stringify(karmaConfig, null, 2)}`);
-    this.server = new karma.Server(karmaConfig, function (exitCode) {
-      process.exit(exitCode);
-    });
-
-    this.listenToBrowserStarted();
+    this.starter = new ProjectStarter(this.loadSetup(options).project);
+    this.cleanRun();
     this.listenToRunComplete();
     this.listenToSpecComplete();
     this.listenToCoverage();
-    this.listenToBrowserError();
+    this.listenToError();
+  }
 
-    this.server.start();
+  init(): Promise<void> {
+    return new Promise((res, rej) => {
+      StrykerReporter.instance.once('browsers_ready', res);
+      this.starter.start()
+        .then(() => { /*noop*/ })
+        .catch(rej);
+    });
   }
 
-  private readConfig(options: RunnerOptions): ConfigOptions {
-    return Object.assign({}, new KarmaConfigReader(options.strykerOptions[KARMA_CONFIG_FILE]).read(), options.strykerOptions[KARMA_CONFIG]);
+  async run({ testHooks }: { testHooks?: string }): Promise<RunResult> {
+    this.testHooksMiddleware.currentTestHooks = testHooks || '';
+    if (this.currentRunStatus !== RunStatus.Error) {
+      // Only run when there was no compile error
+      // An compile error can happen in case of angular-cli
+      await this.runServer();
+    }
+    const runResult = this.collectRunResult();
+    this.cleanRun();
+    return runResult;
   }
 
-  init(): Promise<void> {
-    return this.serverStartedPromise;
+  private loadSetup(settings: RunnerOptions): StrykerKarmaSetup {
```

Comment:
Would it be possible to split this function, it does three things now:

- load the karma config.
- Inform the user their karma configuration is deprecated.
- set globals.